### ISSUE TITLE
MinN/MaxN implementation diff: null handling

### ISSUE THREAD

**Issue by xumingming** [facebookincubator/velox#6589]
### Bug description

Presto test: `TestLongMaxNAggregation::testMixedNullAndNonNullPositions`, which test the max agg function using Blocks, semantically the test equals to the following sql:

```sql
SELECT max(c0, c1)
FROM (
  VALUES 
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '0', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '1', BIGINT '3'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '2', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '3', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '4', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '5', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '6', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '7', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '8', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '9', BIGINT '2')
) AS tmp (c0, c1)
```

In Presto, the query will return `[9, 8]` which is the largest two numbers of the first column, while Velox will throw exception due to the following check(Which is reasonable I think):

```c++
  void checkAndSetN(DecodedVector& decodedN, vector_size_t row) {
    VELOX_USER_CHECK(
        !decodedN.isNullAt(row),
        "second argument of max/min must be a positive integer");
    const auto newN = decodedN.valueAt<int64_t>(row);
    VELOX_USER_CHECK_GT(
        newN, 0, "second argument of max/min must be a positive integer");

    if (n) {
      VELOX_USER_CHECK_EQ(
          newN,
          n,
          "second argument of max/min must be a constant for all rows in a group");
    } else {
      n = newN;
    }
  }
```

My question here is: do we have principle about how to handle these diffs? Especially when Velox's implementation is "more correct"?

### System information

N/A

### Relevant logs

```bash
N/A
```

**Comment by mbasmanova** [comment]
@xumingming This looks like a bug in Velox. min and max functions should ignore null inputs, not throw.

A simpler repro would be

```
SELECT max(c0, n)
FROM (
  VALUES row(1, null), row(2, 2)
) AS t(c0, n)
```

**Comment by mbasmanova** [comment]
CC: @duanmeng @kagamiori @kgpai 

@xumingming Would you like to help fix this?

**Comment by xumingming** [comment]
@mbasmanova I'd be happy to fix it. One thing I want to discuss is that how do we determine it is a bug or just an implementation choice? It seems pretty reasonable to require all the `N` param of min/max to be the same at first glance, is there any standard(Some SQL Spec?) I can refer to?

**Comment by mbasmanova** [comment]
Presto functions in Velox must match Presto semantics.

Similarly, Spark functions must match Spark semantics.

**Comment by mbasmanova** [comment]
> @mbasmanova Sounds reasonable. Presto's checking is pretty loose sometimes though, have you noticed there is a `(BIGINT '1', BIGINT '3')` in the SQL above?

Sorry, I do not understand this question. Would you elaborate a bit?

**Comment by xumingming** [comment]
The intention of the sql here is to calculate `max(c0, 2)`

```sql
SELECT max(c0, c1)
FROM (
  VALUES 
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '0', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '1', BIGINT '3'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '2', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '3', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '4', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '5', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '6', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '7', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '8', BIGINT '2'),
    (cast(NULL as BIGINT), cast(NULL as BIGINT)),
    (BIGINT '9', BIGINT '2')
) AS tmp (c0, c1)
```

But there is a `3` among other `2`s.

**Comment by mbasmanova** [comment]
@xumingming I see. This is a gap in Presto's implementation. Consider creating a GitHub issue in prestodb repo.

**Comment by xumingming** [comment]
Yeah, created an issue: https://github.com/prestodb/presto/issues/20906

**Comment by xumingming** [comment]
min_by/max_by has similar issue, Presto would ignore the nulls for the param `N`:

```bash
presto> SELECT max_by(c0, c1, c2)
     -> FROM (
     ->   VALUES
     ->     (cast(NULL as BIGINT), cast(NULL as BIGINT), cast(NULL as BIGINT)),
     ->     (BIGINT '0', BIGINT '10', BIGINT '2'),
     ->     (cast(NULL as BIGINT), cast(NULL as BIGINT), cast(NULL as BIGINT)),
     ->     (BIGINT '1', BIGINT '9', BIGINT '3'),
     ->     (cast(NULL as BIGINT), cast(NULL as BIGINT), cast(NULL as BIGINT)),
     ->     (BIGINT '2', BIGINT '8', BIGINT '2'),
     ->     (cast(NULL as BIGINT), cast(NULL as BIGINT), cast(NULL as BIGINT)),
     ->     (BIGINT '3', BIGINT '7', BIGINT '2'),
     ->     (cast(NULL as BIGINT), cast(NULL as BIGINT), cast(NULL as BIGINT)),
     ->     (BIGINT '4', BIGINT '6', BIGINT '2'),
     ->     (cast(NULL as BIGINT), cast(NULL as BIGINT), cast(NULL as BIGINT)),
     ->     (BIGINT '5', BIGINT '5', BIGINT '2'),
     ->     (cast(NULL as BIGINT), cast(NULL as BIGINT), cast(NULL as BIGINT)),
     ->     (BIGINT '6', BIGINT '4', BIGINT '2'),
     ->     (cast(NULL as BIGINT), cast(NULL as BIGINT), cast(NULL as BIGINT)),
     ->     (BIGINT '7', BIGINT '3', BIGINT '2'),
     ->     (cast(NULL as BIGINT), cast(NULL as BIGINT), cast(NULL as BIGINT)),
     ->     (BIGINT '8', BIGINT '2', BIGINT '2'),
     ->     (cast(NULL as BIGINT), cast(NULL as BIGINT), cast(NULL as BIGINT)),
     ->     (BIGINT '9', BIGINT '1', BIGINT '2')
     -> ) AS tmp (c0, c1, c2)
     -> ;
 _col0
--------
 [0, 1]
(1 row)
```

Velox don't like NULL in `N`:

```
// MinMaxByAggregates.cpp
void checkAndSetN(DecodedVector& decodedN, vector_size_t row) {
  VELOX_USER_CHECK(
      !decodedN.isNullAt(row),
      "third argument of max_by/min_by must be a positive integer");
```

**Comment by xumingming** [comment]
Fixed by #6667
